### PR TITLE
✨Enhancement/detect user logged out

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

if a user closes a tab opened on a study, and has another tab showing the dashboard, the project will not automatically unlock unless the user refreshes the remaining tab.
This PR shall solve this by sending a websocket event to all tabs a user has opened.

## Related issue/s
fixes [#431](https://github.com/ITISFoundation/osparc-issues/issues/431)

### CODE REVIEW

**Review by GitHK — APPROVED** [review]
👍 looks nice

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 1]
```diff
@@ -186,14 +194,34 @@ async def remove_services_and_data():
     fire_and_forget_task(remove_services_and_data())
 
 
-## PROJECT NODES -----------------------------------------------------
+@observe(event="SIGNAL_USER_DISCONNECTED")
+async def user_disconnected(
+    user_id: int, client_session_id: str, app: web.Application
+) -> None:
+    # check if there is a project resource
+    with managed_resource(user_id, client_session_id, app) as rt:
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:202]: ``user_disconnected`` is a slot to a signal ``SIGNAL_USER_DISCONNECTED`` emitted from a ``managed_resource`` context manager as well. Question: does this context manager supports nested calls or should you pass the caller's context to this function?

File: services/web/server/src/simcore_service_webserver/socketio/handlers.py [hunk 2]
```diff
@@ -123,8 +123,7 @@ async def on_user_logout(
     with managed_resource(user_id, client_session_id, app) as rt:
         # start by disconnecting this client if possible
         if client_session_id:
-            socket_id = await rt.get_socket_id()
-            if socket_id:
+            if socket_id := await rt.get_socket_id():
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/socketio/handlers.py:126]: hahah! fancy py3.8!

File: services/web/server/src/simcore_service_webserver/socketio/handlers.py [hunk 3]
```diff
@@ -153,6 +152,9 @@ async def disconnect(sid: str, app: web.Application) -> None:
             with managed_resource(user_id, client_session_id, app) as rt:
                 log.debug("client %s disconnected from room %s", user_id, sid)
                 await rt.remove_socket_id()
+            # signal same user other clients if available
+            await emit("SIGNAL_USER_DISCONNECTED", user_id, client_session_id, app)
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/socketio/handlers.py:156]: shall these signals be fire & forget?

**Review by pcrespov — APPROVED** [review]

**Review by odeimaiz — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 1 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:202]: it's not nested.

File: services/web/server/src/simcore_service_webserver/socketio/handlers.py (hunk 3 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/socketio/handlers.py:156]: the message posting to the websocket is I think

**Pull request merged**